src/lib/state/DY.Lib.State.Tagged.fst

```diff
@@ -87,27 +92,24 @@ val mk_global_local_bytes_state_predicate_correct: invs:protocol_invariants -> l
   )
   (ensures for_allP (has_local_bytes_state_predicate invs) lpreds)
 let mk_global_local_bytes_state_predicate_correct invs lpreds =
+  no_repeats_p_implies_all_disjoint (List.Tot.map fst lpreds);
   for_allP_eq (has_local_bytes_state_predicate invs) lpreds;
-  FStar.Classical.forall_intro_2 (FStar.Classical.move_requires_2 (mk_global_pred_correct split_local_bytes_state_predicate_func lpreds))
+  FStar.Classical.forall_intro_2 (FStar.Classical.move_requires_2 (mk_global_fun_correct split_local_bytes_state_predicate_func lpreds))
 
 val mk_global_local_bytes_state_predicate_later:
   cinvs:crypto_invariants -> lpreds:list (string & local_bytes_state_predicate) ->
   tr1:trace -> tr2:trace -> prin:principal -> sess_id:state_id -> full_content:bytes -> Lemma
   (requires mk_global_local_bytes_state_predicate lpreds tr1 prin sess_id full_content /\ tr1 <$ tr2)
   (ensures mk_global_local_bytes_state_predicate lpreds tr2 prin sess_id full_content)
 let mk_global_local_bytes_state_predicate_later cinvs lpreds tr1 tr2 prin sess_id full_content =
-  mk_global_pred_eq split_local_bytes_state_predicate_func lpreds (tr1, prin, sess_id, full_content);
-  eliminate exists tag lpred raw_data.
-    List.Tot.memP (tag, lpred) lpreds /\
-    split_local_bytes_state_predicate_func.apply_local_pred lpred raw_data /\
-    split_local_bytes_state_predicate_func.decode_tagged_data (tr1, prin, sess_id, full_content) == Some (split_local_bytes_state_predicate_func.encode_tag tag, raw_data)
-  returns mk_global_local_bytes_state_predicate lpreds tr2 prin sess_id full_content
-  with _. (
-    let Some (_, (_, _, _, content)) = split_local_bytes_state_predicate_func.decode_tagged_data (tr1, prin, sess_id, full_content) in
-    lpred.pred_later tr1 tr2 prin sess_id content;
-    mk_global_pred_eq split_local_bytes_state_predicate_func lpreds (tr2, prin, sess_id, full_content);
-    assert(split_local_bytes_state_predicate_func.apply_local_pred lpred (tr2, prin, sess_id, content))
-  )
+  mk_global_fun_eq split_local_bytes_state_predicate_func lpreds (tr1, prin, sess_id, full_content);
+  mk_global_fun_eq split_local_bytes_state_predicate_func lpreds (tr2, prin, sess_id, full_content);
+  introduce forall lpred content. split_local_bytes_state_predicate_func.apply_local_fun lpred (tr1, prin, sess_id, content) ==> split_local_bytes_state_predicate_func.apply_local_fun lpred (tr2, prin, sess_id, content) with (
+    introduce _ ==> _ with _. lpred.pred_later tr1 tr2 prin sess_id content
+  );
+  match split_local_bytes_state_predicate_func.decode_tagged_data (tr1, prin, sess_id, full_content) with
+  | Some (_, (_, _, _, _)) -> ()
```

Comment:
i think this pattern can be simplified to `Some (_,_)` .
(but interestingly not to just `Some _` ?!)

Reply:
Indeed, I think we would need to increase `ifuel` to have `Some _` (this is probably explained in FStarLang/FStar#3076).
I would prefer keeping the general structure, as some kind of documentation on the type of thing being matched. I think I didn't put the names here because they would mostly be the same as variables in the context (e.g. `prin` or `sess_id`), but we could write them to make things more explicit.

Reply:
If we increase ifuel to 2, we can get rid of the whole match. Does anything speak against doing that?

Reply:
I did that, in general I prefer keeping fuel / ifuel as low as possible, but I guess it's okay here because the proof is easy enough for the SMT